core/src/km_core_state_api.cpp

```diff
@@ -363,4 +364,32 @@ km_core_cp * km_core_state_context_debug(
   result[s.size()] = 0;
 
   return result;
-}
\ No newline at end of file
+}
+
+static bool
+state_has_action_type(km_core_state *state, uint8_t type) {
+  return std::any_of(
+      state->actions().begin(), state->actions().end(),
+        [type](const km::core::action &a) { return a.type == type; });
+}
+
+bool
+state_should_invalidate_context(km_core_state *state,
+                     km_core_virtual_key vk,
+                     uint16_t modifier_state,
+                     uint8_t is_key_down,
+                     uint16_t _kmn_unused(event_flags)) {
+  // if emit_keystroke is present, check if a context reset is needed
+  if (state_has_action_type(state, KM_CORE_IT_EMIT_KEYSTROKE)) {
+    if (vk == KM_CORE_VKEY_BKSP && state->context().empty()) {
+        return true; // context is empty - so pass back
+    } else if (is_key_down &&
+              // either a ctrl or alt modifier
+                ((modifier_state & (K_CTRLFLAG | K_ALTFLAG | LCTRLFLAG | RCTRLFLAG | LALTFLAG | RALTFLAG))
+              // or a frame key
+              || vkey_to_contextreset[vk])) {
```

Comment:
Should verify that vk is in-range; this may come back to bite us with touch otherwise when we have vk>255

Reply:
done